File: src/main/scala/org/jinou/http/SimpleHttpServer.scala
/**
 * [Copyright]
 * @author oujinliang
 * Aug 23, 2013 4:16:46 PM
 */
package org.jinou.http

import com.sun.net.httpserver._
import java.io._
import java.net.InetSocketAddress
import java.lang.reflect.Method

/**
 *
 */
case class SimpleHttpServer(port: Int, classes: Class[_]*) {
    require(port > 0, "port should be greater than 0")
    require(classes.length > 0, "No handler class specified")
    var started = false
    def start() {
        if (started) {
            println("started")
            return
        }
        val server = HttpServer.create(new InetSocketAddress(port), 0)
        server.createContext("/", new MySimpleHandler(classes.toArray))
        server.setExecutor(null) 
        server.start()
        started = true
    }
}

case class HttpMethodPair(method: Method, params: Seq[(String, Class[_])], httpType: String, rule: RouterRule) extends RouterRuleProvider

object Converters {
    val intConverter = (str: String) => Int.box(str.toInt)
    val longConverter = (str: String) => Long.box(str.toLong)
    val floatConverter = (str: String) => Float.box(str.toFloat)
    val doubleConverter = (str: String) => Double.box(str.toDouble)
    val booleanConverter = (str: String) => Boolean.box(str.toBoolean)
    val defaultConverter = (str: String) => str
    
    val converterMap: Map[Class[_], String => AnyRef] = Map(
        classOf[Int] -> intConverter, classOf[Integer] -> intConverter,
        classOf[Long] -> longConverter, classOf[java.lang.Long] -> longConverter,
        classOf[Float] -> floatConverter, classOf[java.lang.Float] -> floatConverter,
        classOf[Double] -> doubleConverter, classOf[java.lang.Double] -> doubleConverter,
        classOf[Boolean] -> booleanConverter, classOf[java.lang.Boolean] -> booleanConverter
    )
    def convert(cls: Class[_], str: String) = converterMap.getOrElse(cls, defaultConverter)(str) 
}
private class MySimpleHandler(classes: Array[Class[_]]) extends HttpHandler {
    import RouterRule._
    
    private val expectedMethods = Array(classOf[GET], classOf[POST])
    private val httpMethodRules = classes map (getHttpHandlerMethods) flatten

    
    @throws(classOf[IOException])
    override def handle(t: HttpExchange) {
        val (code, response) = 
            try { 
                call(t) 
            } catch { 
                case e: Exception => (404, e.toString + "\n" + e.getStackTraceString) 
            }
        val responseStr = if (response == null) "" else response.toString
        t.sendResponseHeaders(code, responseStr.length());
        val os = t.getResponseBody();
        os.write(responseStr.getBytes());
        os.close();
    }
    
    private def call(t: HttpExchange): (Int, Any) = {
       val uri = t.getRequestURI() 
       check(httpMethodRules, uri.toString) map { case (rule, map) =>
           val pair = rule.asInstanceOf[HttpMethodPair]
           val args = (pair.params map { case (name, cls) => Converters.convert(cls, map(name)) }) toArray 

           // TODO: currently only static method is supported
           val method = pair.method
           val field = method.getDeclaringClass().getField("MODULE$");
           val instance = field.get(null);
           val result = pair.method.invoke(instance, args)

           (200, if (result == null) "null" else result.toString) 
       } getOrElse (200, "Bad request")
    }

    private def getHttpHandlerMethods(clz: Class[_]) = {
        (clz.getDeclaredMethods() map { m =>
            val annotations = m.getAnnotations.filter(a => expectedMethods.contains(a.annotationType()))
            if (annotations.isEmpty) None else Some(m, getParams(m), annotations(0)) 
        }).flatten map { case (method, params, httpMethod) =>
            httpMethod match {
                case get: GET   => HttpMethodPair(method, params, "get", RouterRule(get.value()))
                case post: POST => HttpMethodPair(method, params,"post", RouterRule(post.value()))
                case e          => throw new IllegalStateException("unknown method " + e)            
            }  
        }
    }
    
    private def getParams(method: Method) = {
        method.getParameterAnnotations().zip(method.getParameterTypes()) map { case (annotations, paramType) =>
            val params = annotations.filter(a => a.annotationType() == classOf[Param])
            require(params.length > 0, "missing param annotation in method " + method)
            (params(0).asInstanceOf[Param].value(), paramType)
        }
    }
}

File: java/src/main/java/com/xiaomi/http/SimpleHttpServer.java
/**
 * [Copyright] 
 * @author oujinliang
 * Aug 27, 2013 1:27:44 PM
 */
package com.xiaomi.http;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 *
 */
@SuppressWarnings("restriction")
public class SimpleHttpServer {
    private int port;
    private boolean started = false;
    private Class<?>[] handlerClasses;

    public SimpleHttpServer(int port, Class<?>... classes) {
        this.port = port;
        this.handlerClasses = classes;
    }
    
    public void start() throws IOException {
        if (started) {
            return;
        }
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        server.createContext("/", new MySimpleHandler(handlerClasses));
        server.setExecutor(null);
        server.start();
        started = true;
    }
    
    
    static class MySimpleHandler implements HttpHandler {
        private static List<Class<?>> ExpectedMethods = Arrays.asList(new Class<?>[] {GET.class, POST.class});
        private List<HttpMethodArgs> httpMethodArgs = new ArrayList<HttpMethodArgs>();
        
        MySimpleHandler(Class<?>[] classes) {
            for(Class<?> clz : classes) {
                httpMethodArgs.addAll(getHttpMethodArgs(clz));
            }
        }
        
        @Override
        public void handle(HttpExchange he) throws IOException {
            int code = 200;
            String response = null;
            try {
                response = call(he);
            } catch (Exception e) {
                code = 404;
                StringWriter errors = new StringWriter();
                e.printStackTrace(new PrintWriter(errors));
                response = errors.toString();
            }
            
            he.sendResponseHeaders(code, response.length());
            OutputStream os = he.getResponseBody();
            os.write(response.getBytes());
            os.close();
        }
        
        private String call(HttpExchange e) throws Exception {
            String url = e.getRequestURI().toString();
            for (HttpMethodArgs methodArgs : httpMethodArgs) {
                Map<String, String> map = methodArgs.rule.check(url);
                if (map == null) {
                    continue;
                }
                
                Object[] args = getArgs(map, methodArgs.params);
                // TODO: currently only static method is supported
                Object result = methodArgs.method.invoke(null, args);
                return result == null ? "null" : result.toString();
            }
            
            throw new IllegalArgumentException("Bad request");
        }

        static Object[] getArgs(Map<String, String> map, List<MethodParam> params) {
            List<Object> result = new ArrayList<Object>(params.size());
            for (MethodParam param : params) {
                result.add(Converters.convert(param.clz, map.get(param.name)));
            }
            return result.toArray();
        }
        
        static List<HttpMethodArgs> getHttpMethodArgs(Class<?> clz) {
            List<HttpMethodArgs> result = new ArrayList<HttpMethodArgs>();
            for(Method m : clz.getDeclaredMethods()) {
                Annotation a = getHttpMethodAnnotation(m);
                if (a == null) {
                    continue;
                }
                
                List<MethodParam> params = getParam(m);
                if (a instanceof GET) {
                    result.add(new HttpMethodArgs(m, params, "get", new RouterRule(((GET)a).value())));
                } else if (a instanceof POST) {
                    result.add(new HttpMethodArgs(m, params, "post", new RouterRule(((POST)a).value())));
                } else {
                    throw new IllegalArgumentException("unknown method");
                }
            }
            
            return result;
        }
            
        static Annotation getHttpMethodAnnotation(Method m) {
            for (Annotation a : m.getAnnotations()) {
                if (ExpectedMethods.contains(a.annotationType())) {
                    return a;
                }
            }
            return null;
        }
        
        static List<MethodParam> getParam(Method m) {
            List<MethodParam> result = new ArrayList<MethodParam>();
            Annotation[][] paramAnnotations = m.getParameterAnnotations();
            Class<?>[] paramTypes = m.getParameterTypes();
            for (int i = 0; i < paramTypes.length; ++i) {
                Param param = null;
                for(Annotation a : paramAnnotations[i]) {
                    if (a.annotationType().equals(Param.class)) {
                        param = (Param) a;
                        break;
                    }
                }
                if (param == null) {
                    throw new IllegalArgumentException("minssing param annotation in method " + m);
                }
                
                result.add(new MethodParam(param.value(), paramTypes[i]));
            }
            
            return result;
        }
    }
    
    static class HttpMethodArgs {
        final Method method;
        final List<MethodParam> params;
        final String httpType;
        final RouterRule rule;
        
        HttpMethodArgs(Method method, List<MethodParam> params, String httpType, RouterRule rule) {
            this.method = method;
            this.params = params;
            this.httpType = httpType;
            this.rule = rule;
        }
    }
    
    static class MethodParam {
        final String name;
        final Class<?> clz;
        MethodParam(String name, Class<?> clz) {
            this.name = name;
            this.clz = clz;
        }
    }
}
